pyro/contrib/tracking/dynamic_models.py

```diff
@@ -0,0 +1,404 @@
+        '''
+        return self._dimension_pv
+
+    @property
+    def num_process_noise_parameters(self):
+        '''
+        Process noise parameters space dimension access.
+        '''
+        return self._num_process_noise_parameters
+
+    @abstractmethod
+    def __call__(self, x, dt, do_normalization=True):
+        '''
+        Integrate native state ``x`` over time interval ``dt``.
+
+        :param x: current native state. If the DynamicModel is non-differentiable,
+              be sure to handle the case of ``x`` being augmented with process
+              noise parameters.
+        :param dt: time interval to integrate over.
+        :param do_normalization: whether to perform normalization on output, e.g.,
+              mod'ing angles into an interval.
+        :return: Native state x integrated dt into the future.
+        '''
+        raise NotImplementedError
+
+    def geodesic_difference(self, x1, x0):
+        '''
+        Compute and return the geodesic difference between 2 native states.
+        This is a generalization of the Euclidean operation ``x1 - x0``.
+
+        :param x1: native state.
+        :param x0: native state.
+        :return: Geodesic difference between native states ``x1`` and ``x2``.
+        '''
+        return x1 - x0  # Default to Euclidean behavior.
+
+    @abstractmethod
+    def mean2pv(self, x):
+        '''
+        Compute and return PV state from native state. Useful for combining
+        state estimates of different types in IMM (Interacting Multiple Model)
+        filtering.
+
+        .. warning:: For efficiency, may return a reference to the input.
+            Deepcopy as necessary to prevent unexpected changes.
+
+        :param x: native state estimate mean.
+        :return: PV state estimate mean.
+        '''
+        raise NotImplementedError
+
+    @abstractmethod
+    def cov2pv(self, P):
+        '''
+        Compute and return PV covariance from native covariance. Useful for
+        combining state estimates of different types in IMM (Interacting
+        Multiple Model) filtering.
+
+        .. warning:: For efficiency, may return a reference to the input.
+            Deepcopy as necessary to prevent unexpected changes.
+
+        :param P: native state estimate covariance.
+        :return: PV state estimate covariance.
+        '''
+        raise NotImplementedError
+
+    @abstractmethod
+    def process_noise_cov(self, dt=0.):
+        '''
+        Compute and return process noise covariance (Q).
+
+        :param dt: time interval to integrate over.
+        :return: Read-only covariance (Q). For a DifferentiableDynamicModel, this is
+            the covariance of the native state ``x`` resulting from stochastic
+            integration (for use with EKF). Otherwise, it is the covariance
+            directly of the process noise parameters (for use with UKF).
+        '''
+        raise NotImplementedError
+
+    def sample_process_noise(self, dt=0.):
```

Comment:
It would be nice to split this up for use in probabilistic programs:
```py
def process_noise_dist(self, dt=0.):
    return dist.MultivariateNormal(...)

def sample_process_noise(self, dt=0):
   return self.process_noise_dist(dt=dt).sample()
```